### PR TITLE
Allows DALI to dlopen dependent CUDA toolkit libraries: NPP, cuFFT, nvJPEG and nvJPEG2000

### PR DESCRIPTION
**Pull request by JanuszL**
- allows DALI to link dynamically with CUDA toolkit libraries instead of static linking them
- adds dynamic load wrappers for NPP, cuFFT, nvJPEG and nvJPEG2000
- adds WITH_DYNAMIC_CUDA_LIBS CMake build flag that enables dynamic loading of CUDA libraries

Signed-off-by: Janusz Lisiecki <jlisiecki@nvidia.com>

## Description
- [ ] **Bug fix** (*non-breaking change which fixes an issue*)
- [x] **New feature** (*non-breaking change which adds functionality*)
- [ ] **Breaking change** (*fix or feature that would cause existing functionality to not work as expected*)
- [ ] **Refactoring** (*Redesign of existing code that doesn't affect functionality*)
- [ ] **Other** (*e.g. Documentation, Tests, Configuration*)

### What happened in this PR
- allows DALI to link dynamically with CUDA toolkit libraries instead of static linking them
- adds dynamic load wrappers for NPP, cuFFT, nvJPEG and nvJPEG2000
- adds WITH_DYNAMIC_CUDA_LIBS CMake build flag that enables dynamic loading of CUDA libraries

#### Additional information
- Affected modules and functionalities:
  - cmake build system
  - npp, cuFFT, nvJPEG and nvJPEG2000 helpers

- Key points relevant for the review:
  - NA

### Tests
- [X] Existing tests apply
- [ ] New tests added
  - [ ] Python tests
  - [ ] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [X] Existing documentation applies
- [ ] Documentation updated
  - [ ] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [ ] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [X] N/A

**REQ IDs**: N/A

**JIRA TASK**: N/A
DALI-2469

### CODE REVIEW

**Comment by JanuszL** [comment]
!build

**Comment by JanuszL** [comment]
!build

File: cmake/Dependencies.aarch64-qnx.cmake [hunk 1]
```diff
@@ -39,15 +39,24 @@ CUDA_find_library(CUDART_LIB cudart_static)
...
+
+list(APPEND DALI_LIBS ${CUDA_nppicc_LIBRARY})
 list(APPEND DALI_EXCLUDES libnppicc_static.a)
-list(APPEND DALI_LIBS ${CUDA_nppc_static_LIBRARY})
+list(APPEND DALI_LIBS ${CUDA_nppc_LIBRARY})
 list(APPEND DALI_EXCLUDES libnppc_static.a)
 
-# cuFFT library
-CUDA_find_library(CUDA_cufft_static_LIBRARY cufft_static)
+# cuFFT libraryif (WITH_DYNAMIC_CUDA_LIBS)
```
Comment (stiepan) [inline cmake/Dependencies.aarch64-qnx.cmake:55]: missing newline?

File: cmake/Dependencies.aarch64-qnx.cmake [hunk 2]
```diff
@@ -39,15 +39,24 @@ CUDA_find_library(CUDART_LIB cudart_static)
 list(APPEND DALI_EXCLUDES libcudart_static.a)
 
 # NVIDIA NPP library
-CUDA_find_library(CUDA_nppicc_static_LIBRARY nppicc_static)
-CUDA_find_library(CUDA_nppc_static_LIBRARY nppc_static)
-list(APPEND DALI_LIBS ${CUDA_nppicc_static_LIBRARY})
+if (WITH_DYNAMIC_CUDA_LIBS)
+CUDA_find_library(CUDA_nppic_LIBRARY nppicc)
```
Comment (stiepan) [inline cmake/Dependencies.aarch64-qnx.cmake:43]: nitpick: missing indentation

File: cmake/Dependencies.cmake [hunk 3]
```diff
@@ -53,16 +58,20 @@ if (BUILD_NVJPEG2K)
...
 
 # cuFFT library
-CUDA_find_library(CUDA_cufft_static_LIBRARY cufft_static)
-list(APPEND DALI_EXCLUDES libcufft_static.a)
+if (NOT WITH_DYNAMIC_CUDA_LIBS)
+  CUDA_find_library(CUDA_cufft_LIBRARY cufft_static)
+  list(APPEND DALI_EXCLUDES libcufft_static.a)
+endif (NOT WITH_DYNAMIC_CUDA_LIBS)
 
 # CULIBOS needed when using static CUDA libs
```
Comment (stiepan) [inline cmake/Dependencies.cmake:76]: So is it needed in dynamic linking case?

File: cmake/Dependencies.aarch64-qnx.cmake (hunk 2 above)
Comment (JanuszL) [reply cmake/Dependencies.aarch64-qnx.cmake:43]: I forgot to update this file. Fixed.

File: cmake/Dependencies.cmake (hunk 3 above)
Comment (JanuszL) [reply cmake/Dependencies.cmake:76]: Good catch. I need to check it.

Comment (JanuszL) [reply cmake/Dependencies.cmake:76]: ... on the other head dlopen doesn't load other libs so we would need to load them on our own and provide the dynamic wrapper as well. I'm not sure if we want to maintain such entity we don't use directly.
@mzient - what do you think?

File: cmake/Dependencies.aarch64-qnx.cmake (hunk 1 above)
Comment (stiepan) [reply cmake/Dependencies.aarch64-qnx.cmake:51]: Nitpick: This unnecessary if stayed in the comment.

File: cmake/Dependencies.cmake (hunk 3 above)
Comment (JanuszL) [reply cmake/Dependencies.cmake:76]: It seems to work, and dynamic libs doesn't require CULIBOS. Let me run the CI to be sure.

File: CMakeLists.txt [hunk 4]
```diff
@@ -39,6 +39,11 @@ check_cxx_compiler_flag(-fopenmp-simd CXX_HAVE_OMP_SIMD)
 # Build options
 option(BUILD_DALI_NODEPS "Disable components that require extra external libraries to be present in the system. Effectively, it builds only the DALI core and kernel libraries")
 option(LINK_DRIVER "Links directly with libcuda.so instead of dlopen it at runtime" OFF)
+option(WITH_DYNAMIC_CUDA_LIBS "Links CUDA libraires dynamically (NPP, nvJPEG, cuFFT)" OFF)
+
+if (WITH_DYNAMIC_CUDA_LIBS)
+  add_definitions(-DWITH_DYNAMIC_CUDA_LIBS=1)
+endif(WITH_DYNAMIC_CUDA_LIBS)
```
Comment (szalpal) [inline CMakeLists.txt:46]:
```suggestion
endif()
```
I believe it's not necessary anymore for a few years ;) Same thing applies to the rest of `ifs` below.

File: dali/kernels/CMakeLists.txt [hunk 5]
```diff
@@ -38,7 +38,12 @@ adjust_source_file_language_property("${DALI_KERNEL_SRCS}")
 set_source_files_properties("dummy.cu" PROPERTIES LANGUAGE CUDA)
 add_library(dali_kernels ${LIBTYPE} ${DALI_KERNEL_SRCS})
 target_link_libraries(dali_kernels PUBLIC dali_core)
-target_link_libraries(dali_kernels PRIVATE ${CUDA_cufft_static_LIBRARY})
+if (WITH_DYNAMIC_CUDA_LIBS)
+  target_link_libraries(dali_kernels PRIVATE dynlink_cufft)
+else()
+  target_link_libraries(dali_kernels PRIVATE ${CUDA_cufft_LIBRARY})
+  target_link_libraries(dali_kernels PRIVATE "-Wl,--exclude-libs,${CUDA_cufft_LIBRARY}")
```
Comment (szalpal) [inline dali/kernels/CMakeLists.txt:45]: I'm a little bit lost at these 2 lines. Firstly we link `CUDA_cufft_LIBRARY` and then we link it, passing `exclude-libs` option to the linker?

File: tools/stub_generator/nvjpeg.json [hunk 6]
```diff
@@ -0,0 +1,44 @@
+{
+   "extra_include":[
+      "<nvjpeg.h>"
+   ],
+   "return_type":"nvjpegStatus_t",
+   "calling_conv":"NVJPEGAPI",
+   "not_found_error":"NVJPEG_STATUS_IMPLEMENTATION_NOT_SUPPORTED",
+"functions": {
+      "nvjpegCreate": {},
+      "nvjpegDestroy": {},
```
Comment (szalpal) [inline tools/stub_generator/nvjpeg.json:10]: Do I understand correctly, that whenever we use some unlisted here nvjpeg function (or nvjpeg adds a function to the API), we should update this file?

File: CMakeLists.txt [hunk 7]
```diff
@@ -39,6 +39,11 @@ check_cxx_compiler_flag(-fopenmp-simd CXX_HAVE_OMP_SIMD)
 # Build options
 option(BUILD_DALI_NODEPS "Disable components that require extra external libraries to be present in the system. Effectively, it builds only the DALI core and kernel libraries")
 option(LINK_DRIVER "Links directly with libcuda.so instead of dlopen it at runtime" OFF)
+option(WITH_DYNAMIC_CUDA_LIBS "Links CUDA libraires dynamically (NPP, nvJPEG, cuFFT)" OFF)
```
Comment (szalpal) [inline CMakeLists.txt:42]: How does `WITH_DYNAMIC_CUDA_LIBS` option relate to `LINK_CUDA_DYNAMICALLY`? Can these two options be opposite (i.e. `WITH_DYNAMIC_CUDA_LIBS xor LINK_CUDA_DYNAMICALLY == 1`)?

**Review by szalpal — COMMENTED** [review]
I asked some questions in the comments

File: tools/stub_generator/nvjpeg.json (hunk 6 above)
Comment (JanuszL) [reply tools/stub_generator/nvjpeg.json:10]: Yes we should, or if someone really wants can list all the functions nvJPEG has. I decided to do it lazily.

File: CMakeLists.txt (hunk 7 above)
Comment (JanuszL) [reply CMakeLists.txt:42]: LINK_CUDA_DYNAMICALLY is about libcuda.so (driver) while WITH_DYNAMIC_CUDA_LIBS is about cuda toolkit. I will rename WITH_DYNAMIC_CUDA_LIBS to WITH_DYNAMIC_CUDA_TOOLKIT.

File: dali/kernels/CMakeLists.txt (hunk 5 above)
Comment (JanuszL) [reply dali/kernels/CMakeLists.txt:45]: _Specifies a list of archive libraries from which symbols should not be automatically exported. The library names may be delimited by commas or colons. Specifying --exclude-libs ALL excludes symbols in all archive libraries from automatic export. This option is available only for the i386 PE targeted port of the linker and for ELF targeted ports. For i386 PE, symbols explicitly listed in a .def file are still exported, regardless of this option. For ELF targeted ports, symbols affected by this option will be treated as hidden._

File: CMakeLists.txt (hunk 4 above)
Comment (JanuszL) [reply CMakeLists.txt:46]: Done

File: CMakeLists.txt [hunk 8]
```diff
@@ -39,6 +39,11 @@ check_cxx_compiler_flag(-fopenmp-simd CXX_HAVE_OMP_SIMD)
 # Build options
 option(BUILD_DALI_NODEPS "Disable components that require extra external libraries to be present in the system. Effectively, it builds only the DALI core and kernel libraries")
 option(LINK_DRIVER "Links directly with libcuda.so instead of dlopen it at runtime" OFF)
+option(WITH_DYNAMIC_CUDA_TOOLKIT "Links CUDA toolkit libraires dynamically (NPP, nvJPEG, cuFFT)" OFF)
+
+if (WITH_DYNAMIC_CUDA_TOOLKIT)
+  add_definitions(-DWITH_DYNAMIC_CUDA_TOOLKIT=1)
```
Comment (klecki) [inline CMakeLists.txt:45]: Can you consider making it similar to `propagate_option` which exports 1 and 0 valued define in both cases?

**Comment by JanuszL** [comment]
!build

File: cmake/Dependencies.aarch64-qnx.cmake [hunk 9]
```diff
@@ -74,7 +79,7 @@ include(cmake/Dependencies.common.cmake)
 
 ##################################################################
 # Protobuf
-##################################################################
+\##################################################################
```
Comment (stiepan) [inline cmake/Dependencies.aarch64-qnx.cmake:82]: nitpick: stray "\"?

**Review by stiepan — APPROVED** [review]

File: cmake/Dependencies.aarch64-qnx.cmake (hunk 9 above)
Comment (JanuszL) [reply cmake/Dependencies.aarch64-qnx.cmake]: Done

**Review by szalpal — APPROVED** [review]
My only concern is that it's not obvious to update `stub_generator/*.json` files whenever a new lib function will be used. Especially if it's not often, that we'll update it. Someone might loose quite a bit of time and compute trying to figure out, why the CI is not passing. Is it possible to inform the devs somehow about it? Unfortunately, I don't have good idea how to do it...

**Comment by JanuszL** [comment]
> My only concern is that it's not obvious to update `stub_generator/*.json` files whenever a new lib function will be used. Especially if it's not often, that we'll update it. Someone might loose quite a bit of time and compute trying to figure out, why the CI is not passing. Is it possible to inform the devs somehow about it? Unfortunately, I don't have good idea how to do it...

We have been applying this approach to cuda driver, NVDEC and Optical flow for more than a year now without major problems. And I don't expect them to start appearing soon.

**Pull request merged**